server/util.go

```diff
@@ -209,34 +209,37 @@ func interceptConfigs(rootViper *viper.Viper) (*tmcfg.Config, error) {
 		rootViper.SetConfigType("toml")
 		rootViper.SetConfigName("config")
 		rootViper.AddConfigPath(configPath)
+
 		if err := rootViper.ReadInConfig(); err != nil {
-			return nil, fmt.Errorf("failed to read in app.toml: %w", err)
+			return nil, fmt.Errorf("failed to read in %s: %w", tmCfgFile, err)
 		}
 	}
 
-	// Read into the configuration whatever data the viper instance has for it
-	// This may come from the configuration file above but also any of the other sources
-	// viper uses
+	// Read into the configuration whatever data the viper instance has for it.
+	// This may come from the configuration file above but also any of the other
+	// sources viper uses.
 	if err := rootViper.Unmarshal(conf); err != nil {
 		return nil, err
 	}
+
 	conf.SetRoot(rootDir)
 
-	appConfigFilePath := filepath.Join(configPath, "app.toml")
-	if _, err := os.Stat(appConfigFilePath); os.IsNotExist(err) {
+	appCfgFilePath := filepath.Join(configPath, "app.toml")
+	if _, err := os.Stat(appCfgFilePath); os.IsNotExist(err) {
 		appConf, err := config.ParseConfig(rootViper)
 		if err != nil {
 			return nil, fmt.Errorf("failed to parse app.toml: %w", err)
 		}
 
-		config.WriteConfigFile(appConfigFilePath, appConf)
+		config.WriteConfigFile(appCfgFilePath, appConf)
 	}
 
 	rootViper.SetConfigType("toml")
 	rootViper.SetConfigName("app")
-	rootViper.AddConfigPath(configPath)
-	if err := rootViper.ReadInConfig(); err != nil {
-		return nil, fmt.Errorf("failed to read in app.toml: %w", err)
+	rootViper.AddConfigPath(appCfgFilePath)
+
+	if err := rootViper.MergeInConfig(); err != nil {
```

Comment:
This was causing any user value set in Tendermint's config to be wiped out. We need to merge configurations, not overwrite them.